### ISSUE TITLE
Intent definition attributes

### ISSUE THREAD

**Issue by kwridan** [tuist/tuist#3227]
> I'm trying to figure out if there's a way to set target membership options for an `.intentdefinition` file. Would appreciate if you can help with a way around this.
>
> <img width="143" alt="Target_Membership" src="https://user-images.githubusercontent.com/82376303/126837512-393047f5-c951-4064-b415-f3f4e1c9e0e2.png">
> 
_Originally posted by @soaurabhcompass in https://github.com/tuist/tuist/issues/3079#issuecomment-885896451_

**Comment by kwridan** [comment]
Hi @soaurabhcompass I've moved the discussion for this here as it appears to be a separate issue.

Having had a quick look, looks like those options are defined as attributes on the intents definition file

e.g. Selecting private 

```diff
- FBFEC58226AE986B00278DA4 /* Intents.intentdefinition in Sources */ = {isa = PBXBuildFile; fileRef = FBFEC58126AE986B00278DA4 /* Intents.intentdefinition */;  };
+ FBFEC58226AE986B00278DA4 /* Intents.intentdefinition in Sources */ = {isa = PBXBuildFile; fileRef = FBFEC58126AE986B00278DA4 /* Intents.intentdefinition */; settings = {ATTRIBUTES = (private_codegen, ); }; };
```

The mapping appears to be:

- **Public Intent Classes**: _No attributes_ (or `codgen`)
- **Private Intent Classes**: `private_codegen`
- **Project Intent Classes**: `project_codegen`
- **No Generated Classes**: `no_codegen`

There currently isn't a way to specify this sadly, but something that could be added.

**Comment by kwridan** [comment]
A few initial thoughts:

Seeing [`SourceFileGlob`](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/SourceFilesList.swift#L12) already has a `compilersFlag` property, perhaps another property could be added there too. Here are some options:

**Raw attributes**

- `attributes: [String]?`: where the raw attribute (e.g `private_codegen` ) could be defined

example:

```swift
sources: [
   SourceFileGlob("intents/**", attributes: ["private_codegen"])
]
```

This is probably the most powerful / flexible option, but exposes some of the underlying Xcode details which is inconsistent with the rest of the APIs defined by Tuist.

**Raw + typed attributes**

An expansion of the previous option, this time it's a typed attributed and we offer a few predefined attributes alongside the ability to define raw ones. 

- `attributes: [FileAttribute]?`

```swift
struct FileAttribute {
   var value: String
}

extension FileAttribute {
  static var privateCodegen: FileAttribute { FileAttribute(value: "private_codegen") }
  // ...
}
```

example:

```swift
sources: [
   SourceFileGlob("intents/**", attributes: [.privateCodegen])
]
```

This option still exposes the underlying raw types should you need it _(e.g. other attributes Tuist doesn't yet support)_, but also offers some strongly typed predefined options too. 

**Specific intents flags**

Another could be to be more restrictive and specific, and only expose a codegen enum:

- `codegen: FileCodeGen?`

```swift
enum FileCodeGen {
  case `public`
  case `private`
  case project
  case disabled
}
```

example:

```swift
sources: [
   SourceFileGlob("intents/**", codegen: .private)
]
```

I suspect CoreData models have similar attributes. This is more user-friendly and will resolve this specific case. Though does mean more properties would need to be added to `FileSourceGlob` as we discover new ones. Perhaps with some lint checks we could flag misuse of the flag _(e.g. setting code gen attributes for regular swift files)_.

**Comment by fortmarek** [comment]
Thanks @kwridan for the research on this!

I think `specific intents flags` are the most optimal solution. We don't want users to dig through Xcode to get a raw value but rather tell us that something is missing, so we can make it easier and possible to do for _everyone_.  

> I suspect CoreData models have similar attributes.

The `codegen` attribute is also used when you import a `.mlmodel` to your project.

> Perhaps with some lint checks we could flag misuse of the flag

Yeah, that's a good idea - I wonder if we should flip that and add lint checks that the extensions of files added are only of specific types. I believe the set of possible file extensions that can be used is limited - however, it does increase the maintenance burden (e.g. if Apple adds some new file extension for which it makes sense to add this attribute, we'd need to add it as well).

**Comment by pavm035** [comment]
I think `FileAttribute` approach make sense to me since we (developers) knows that possible fixed types offered by `FileAttribute` and on the other side if apple adds new flag then user will not be blocked until we support new flag instead they can use raw one meanwhile we add new one

**Comment by fortmarek** [comment]
> if apple adds new flag then user will not be blocked until we support new flag instead they can use raw one meanwhile we add new one

That is true, although the current solution would lead users asking questions like what is `custom` useful for? If the only reason for adding `custom` is that we won't have to update the number of cases when Apple releases new Xcode, I think it's not good enough of a reason - we should watch Xcode 13 developments and update it when necessary, rather than polluting the API with types that _most_ of the time won't serve any purpose.

**Comment by pavm035** [comment]
Yea that’s valid point, so meanwhile we would like to go with this approach and planning to fork it and modify for our project so how do you recommend to bundle modified version this library and use it in our project? Because this is show stopper for us now since we almost there with migration

**Comment by fortmarek** [comment]
> Yea that’s valid point, so meanwhile we would like to go with this approach and planning to fork it and modify for our project so how do you recommend to bundle modified version this library and use it in our project?

I'd suggest contributing back to the project, I don't think this should be extremely difficult to implement 🙂 I can provide guidance if necessary.

**Comment by pavm035** [comment]
@fortmarek  sure i can work on it please let me know how to get started and need little knowledge on this domain, but yea fix should be pretty easy

**Comment by kwridan** [comment]
> @fortmarek sure i can work on it please let me know how to get started and need little knowledge on this domain, but yea fix should be pretty easy

Awesome, thanks @pavm035! 

Will outline some of the bits / areas that need to be updated to support this.

## Setup

### Tuist Setup

Worth reading through https://docs.tuist.io/contributors/get-started to get Tuist up and running locally. You can experiment with generating one of the [existing fixtures](https://github.com/tuist/tuist/tree/main/projects/tuist/fixtures) from the local version of Tuist to verify it's all set up correctly.

### Sample Fixture

For this particular issue, it'll be really handy to have a sample fixture that can be iterated on. A new simple fixture can be added to the [`projects/tuist/fixtures/`](https://github.com/tuist/tuist/tree/main/projects/tuist/fixtures) directory which contains an iOS app with few intent definition files. 

e.g.

```swift
Project(
    name: "MyProject",
    targets: [
        Target(
            name: "App",
            // ...
            sources: [
                "App/**",
                "Intents/Public.intentdefinition",
                "Intents/Private.intentdefinition",
                "Intents/Project.intentdefinition",
                "Intents/Disabled.intentdefinition",
            ]
        )
    ]
)
```

Tuist should still be able to generate this fixture (albeit without the new API addition just yet)

## Exposing the new public API

Tuist's public manifest APIs (the types in `Project.swift`) are all defined in the [`ProjectDescription`](https://github.com/tuist/tuist/tree/main/Sources/ProjectDescription) target.

For this case:

- A new `FileCodeGen` codable enum types needs to be added
- [`SourceFileGlob`](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/SourceFilesList.swift#L4) needs to be extended to include an optional `codeGen` property
  - Worth adding a doc comment note about this property being applicable to only intent definition and ml models files
- The [`TargetTests`](https://github.com/tuist/tuist/blob/main/Tests/ProjectDescriptionTests/TargetTests.swift#L59) can be extended to include a few more source samples for coverage which include the `codeGen` property set.

The fixture can then be updated and the local version of Tuist should be able to generate it without errors _(even though the new property still does nothing)_

```swift
Project(
    name: "MyProject",
    targets: [
        Target(
            name: "App",
            // ...
            sources: [
                "App/**",
                SourceFileGlob("Intents/Public.intentdefinition", codeGen: .public),
                SourceFileGlob("Intents/Private.intentdefinition", codeGen: .private),
                SourceFileGlob("Intents/Project.intentdefinition", codeGen: .project),
                SourceFileGlob("Intents/Disabled.intentdefinition", codeGen: .disabled),
            ]
        )
    ]
)
```

## Updating the internal Models

### Adding the new internal API

Tuist internally has a similar set of corresponding models within the [`TuistGraph`](https://github.com/tuist/tuist/tree/main/Sources/TuistGraph/Models) target - while similar those often include additional implementation details and are in the expanded form _(e.g. The resolved absolute paths are available here)_

- A new `FileCodeGen` codable enum types needs to be added
- [`SourceFileGlob`](https://github.com/tuist/tuist/blob/main/Sources/TuistGraph/Models/SourceFileGlob.swift) needs to be extended to include an optional `codeGen` property
- [`SourceFile`](https://github.com/tuist/tuist/blob/main/Sources/TuistGraph/Models/SourceFile.swift) needs to be extended to include an optional `codeGen` property

### Mapping between the public and internal API

Tuist has a series of Manifest Mappers that translate the public manifest APIs (from `ProjectDescription`) to the inernal models (`TuistGraph`). Those live in the [`TuistLoader`](https://github.com/tuist/tuist/tree/main/Sources/TuistLoader/Models%2BManifestMappers) target. 

For sources there's a few hoops to get through to get to the final type `SourceFile` - hopefully not too jarring.

- A new `FileCodeGen+ManifestMapper` can be added to translate `ProjectDescription.FileCodeGen` > `TuistGraph.FileCodeGen`
-  [`Target+ManifestMapper`](https://github.com/tuist/tuist/blob/main/Sources/TuistLoader/Models%2BManifestMappers/Target%2BManifestMapper.swift#L171) can be updated to pass along the translated `FileCodeGen`
- [`Target+Core`](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/ModelExtensions/Target+Core.swift#L83) where globs are evaluated would need to pass along the `codeGen` property to the final `SourceFile` type.

## Updating the generator

The Xcode generation lives within the [`TuistGenerator`](https://github.com/tuist/tuist/tree/main/Sources/TuistGenerator) target. Now that `SourceFile` has the `FileCodeGen` property, the generator can translate it to the raw Xcode attributes. 

- `FileCodeGen` and `SourceFile` can be extended compute the raw Xcode attributes
- The [`BuildPhaseGenerator.generateSourcesBuildPhase`](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift#L217) can be updated to add source file attributes the `settings` dictionary if any exist
- The [`BuildPhaseGeneratorTests`](https://github.com/tuist/tuist/blob/main/Tests/TuistGeneratorTests/Generator/BuildPhaseGeneratorTests.swift#L41) can be extended to include a few cases that have sources with `codeGen` set

If all is wired correctly, the sample fixture should now generate and have the correct attributes set for the intent definition files

## Linting

Most of the generation related linting logic lives in a series of [`Linter`](https://github.com/tuist/tuist/tree/main/Sources/TuistGenerator/Linter) components. Here we'd like to warn users in case `codeGen` is defined on unsupported file types.

- [`TargetLinter`](https://github.com/tuist/tuist/blob/main/Sources/TuistGenerator/Linter/TargetLinter.swift) can be extended to include a check to see if any `SourceFile` has `codeGen` defined on any sources files that don't have an extension `.intentdefinition` or `.mlmodel`

e.g.

```swift
private func lintValidSourceFileCodeGenAttributes(target: Target) -> [LintingIssue] {
    let knownSupportedExtensions = [
        "intentdefinition",
        "mlmodel",
    ]
    let unsupportedSourceFileAttributes = target.sources.filter {
        $0.codeGen != nil && !knownSupportedExtensions.contains($0.path.extension ?? "")
    }

    return unsupportedSourceFileAttributes.map {
        .init(
            reason: "Target '\(target.name)' has a source file at path \($0.path) with unsupported `codeGen` attributes. Only \(knownSupportedExtensions.listed()) are known to support this.",
            severity: .warning
        )
    }
}
```

- [`TargetLinterTests`](https://github.com/tuist/tuist/blob/main/Tests/TuistGeneratorTests/Linter/TargetLinterTests.swift) can be extended to test valid and invalid combination.

e.g.

```swift
let target = Target.empty(
    name: "MyTarget",
    sources: [
        SourceFile(path: "/project/Source.swift"),
        SourceFile(path: "/project/Invalid.swift", codeGen: .project),
        SourceFile(path: "/project/Unspecified.intentdefinition"),
        SourceFile(path: "/project/Valid.intentdefinition", codeGen: .private),
        SourceFile(path: "/project/Valid.mlmodel", codeGen: .disabled),
    ]
)
```

## Updating the docs

Finally seeing this introduces a new API, the [manifest docs](https://github.com/tuist/tuist/blob/main/projects/docs/docs/manifests/project.md) can be updated to include reference to the new property. Worth a note here that this is only applicable to a subset of file types.

**Comment by pavm035** [comment]
Got it let me go through and work on it

**Comment by pavm035** [comment]
@kwridan  @fortmarek  here is the PR
https://github.com/tuist/tuist/pull/3448

**Comment by kwridan** [comment]
Resolved via #3448
